app/src/main/java/com/github/se/wanderpals/model/viewmodel/SuggestionsViewModel.kt

```diff
@@ -9,27 +10,106 @@ import kotlinx.coroutines.flow.StateFlow
 import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.launch
 
-open class SuggestionsViewModel(private val suggestionRepository: TripsRepository, tripId: String) :
-    ViewModel() {
+open class SuggestionsViewModel(
+    private val suggestionRepository: TripsRepository?,
+    tripId: String
+) : ViewModel() {
+
+  private val currentLoggedInUId =
+      suggestionRepository
+          ?.uid!! // Get the current logged-in user's ID from the repository instance
+
   // State flow to hold the list of suggestions
   private val _state = MutableStateFlow(emptyList<Suggestion>())
   open val state: StateFlow<List<Suggestion>> = _state
 
   private val _isLoading = MutableStateFlow(true)
   open val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
 
+  // the like status of each suggestion to be held to prevent repeated network calls for the same
+  // item:
+  private val _likedSuggestions = MutableStateFlow<List<String>>(emptyList())
+
   init {
     // Fetch all trips when the ViewModel is initialized
     loadSuggestion(tripId)
   }
 
+  open fun getIsLiked(suggestionId: String): Boolean {
+    return _likedSuggestions.value.contains(suggestionId)
+  }
+
+  open fun getNbrLiked(suggestionId: String): Int {
+    return _state.value.find { it.suggestionId == suggestionId }?.userLikes?.size ?: 0
+  }
+
   /** Fetches all trips from the repository and updates the state flow accordingly. */
   open fun loadSuggestion(tripId: String) {
     viewModelScope.launch {
       _isLoading.value = true
       // Fetch all trips from the repository
-      _state.value = suggestionRepository.getAllSuggestionsFromTrip(tripId)
+      _state.value = suggestionRepository?.getAllSuggestionsFromTrip(tripId)!!
+      Log.d("Fetched Suggestions", _state.value.toString())
+
+      _likedSuggestions.value =
+          _state.value.filter { it.userLikes.contains(currentLoggedInUId) }.map { it.suggestionId }
       _isLoading.value = false
     }
   }
+
+  /**
+   * Toggles the like status of a suggestion and updates the backend and local state accordingly.
+   *
+   * Note: open keyword is used to allow overriding this function in a subclass of
+   * SuggestionsViewModel, namely the MockSuggestionsViewModel class when testing.
+   */
+  open fun toggleLikeSuggestion(tripId: String, suggestion: Suggestion) {

```

Comment:
This function doesn't properly work when disliking then liking again. The number of likes count down correctly but increased twice when liking again: the userId is added twice to the database instead of once. This needs to be fixed